experimental/sev_guest/src/msr.rs

```diff
@@ -174,6 +190,102 @@ pub fn get_cpuid(request: CpuidRequest) -> Result<CpuidResponse> {
     read_msr().try_into()
 }
 
+/// A request for the hypervisor's preferred location for the GHCB page.
+pub struct PreferredGhcbGpaRequest;
+
+impl From<PreferredGhcbGpaRequest> for u64 {
+    fn from(_request: PreferredGhcbGpaRequest) -> Self {
+        0x010
+    }
+}
+
+/// The response containing the preferred location of the GHCB.
+pub struct PreferredGhcbGpaResponse {
+    /// The preferred guest-physical address for the GHCB.
+    pub ghcb_gpa: usize,
+}
+
+impl TryFrom<u64> for PreferredGhcbGpaResponse {
+    type Error = anyhow::Error;
+    fn try_from(msr_value: u64) -> Result<Self> {
+        const PREFERRED_GPA_RESPONSE_INFO: u64 = 0x011;
+        if msr_value & GHCB_INFO_MASK != PREFERRED_GPA_RESPONSE_INFO {
+            bail!("Value is not a valid preferred GHCP GPA response");
+        }
+        let ghcb_gpa = (msr_value & GCHP_DATA_MASK) as usize;
+        Ok(Self { ghcb_gpa })
+    }
+}
+
+/// Requests the hypervisor's preferred location for the GHCB page.
+///
+/// A response of `NO_PREFERRED_GHCB_LOCATION` indicates that the hypervisor does not have a
+/// preferred location.
+///
+/// The guest must validate that the returned guest-physical address is not part of its known
+/// guest-private memory.
+pub fn get_preferred_ghcb_location() -> Result<PreferredGhcbGpaResponse> {
+    let request = PreferredGhcbGpaRequest;
+    write_msr_and_exit(request.into());
+    read_msr().try_into()
+}
+
+/// Request to register a guest-physical address for the GHCB with the hypervisor.
+pub struct RegisterGhcbGpaRequest {
+    ghcb_gpa: u64,
+}
+
+impl RegisterGhcbGpaRequest {
+    pub fn new(ghcb_gpa: usize) -> Result<Self> {
+        let ghcb_gpa = ghcb_gpa as u64;
+        if ghcb_gpa & GHCB_INFO_MASK != 0 {
+            bail!("GHCB must be 4KiB-aligned");
+        }
+        Ok(Self { ghcb_gpa })
+    }
+}
+
+impl From<RegisterGhcbGpaRequest> for u64 {
+    fn from(register_ghcb_gpa_request: RegisterGhcbGpaRequest) -> Self {
+        let mut result = 0x012;
```

Comment:
Would that be `PREFERED_GHCB_GPA_REQUEST` from GHCB page 14/Table 2?

Reply:
This is the GHCBInfo value for "Register GHCB GPA Request" from that table.